File: frontend/src/lib/pathGenerator.ts
import { SKILLS, getSubtopics, shortestLearningOrder } from './skillGraph';
import { parseGoal, recommendCourses } from './recommender';
import { searchVideos, searchPlaylists, getPlaylistVideos, findTopicChannel } from './youtube';
import type {
  Course, GenerationParams, LearningPath, PathStage, PathTask, PathType, Skill,
} from '@/types';

const LEVEL_TO_STAGE: Record<string, number> = { Beginner: 0, Intermediate: 1, Advanced: 2, Expert: 2 };
const STAGE_ORDER = ['foundation', 'core', 'advanced'];
const DOMAIN_COLOR: Record<string, 'amber' | 'iris' | 'sage'> = {
  programming: 'iris', data_science: 'sage', web_dev: 'amber',
  cloud: 'iris', business: 'amber', design: 'sage', custom: 'iris',
};
const DOMAIN_ICON: Record<string, string> = {
  programming: 'Code2', data_science: 'BarChart3', web_dev: 'Atom',
  cloud: 'Cloud', business: 'Compass', design: 'Palette', custom: 'Sparkles',
};

function inferKnownSkills(experienceLevel: string, domainHint: string | null): string[] {
  const stageIdx = LEVEL_TO_STAGE[experienceLevel] ?? 0;
  if (stageIdx === 0 || !domainHint) return [];
  const allowed = new Set(STAGE_ORDER.slice(0, stageIdx));
  return Object.entries(SKILLS)
    .filter(([, meta]) => meta.domain === domainHint && allowed.has(meta.level))
    .map(([id]) => id);
}

function cleanTopicPhrase(goalText: string): string {
  const filler = /\b(i want to|i wanna|learn|study|become a|get a|placement|for|the|please|help me)\b/gi;
  const cleaned = goalText.replace(filler, ' ').replace(/\s+/g, ' ').trim();
  const phrase = cleaned || goalText;
  return phrase.replace(/\b\w/g, (c) => c.toUpperCase());
}

export async function buildTasksForSkill(
  skillId: string,
  label: string,
  subtopics: string[],
  courses: ReturnType<typeof recommendCourses>,
  resourceMode: GenerationParams['resourceMode'],
  apiKey: string | null,
): Promise<PathTask[]> {
  const tasks: PathTask[] = [];

  if (courses.length) {
    const best = courses[0];
    tasks.push({
      id: `${skillId}_course`,
      type: 'course',
      label: `Take the foundational course: ${best.title}`,
      provider: best.provider,
      url: best.url,
      thumbnail: null,
      meta: `⭐${best.rating || 'N/A'} · ${best.is_paid ? 'Paid' : 'Free'} · ${best.level}`,
    });
  }

  let videoTasks: PathTask[] = [];
  let mode = resourceMode;

  if (mode === 'playlist') {
    const playlists = await searchPlaylists(`${label} full course`, apiKey, 1);
    if (playlists.length) {
      const pl = playlists[0];
      const items = await getPlaylistVideos(pl.playlistId, apiKey, Math.max(subtopics.length, 8));
      videoTasks = items.map((v, j) => ({
        id: `${skillId}_pl${j}`,
        type: 'video' as const,
        label: v.title,
        provider: `${pl.channel} (playlist)`,
        url: v.url,
        thumbnail: v.thumbnail,
        meta: `Part ${j + 1} of playlist: ${pl.title}`,
      }));
    }
    if (!videoTasks.length) mode = 'same_channel';
  }

  if (mode === 'same_channel' && !videoTasks.length) {
    const channel = await findTopicChannel(label, apiKey);
    videoTasks = await Promise.all(
      subtopics.map(async (sub, j) => {
        let vids = await searchVideos(`${sub} tutorial`, apiKey, 1, channel?.channelId ?? null);
        if (!vids.length && channel) vids = await searchVideos(`${sub} tutorial`, apiKey, 1, null);
        const v = vids[0];
        return {
          id: `${skillId}_sub${j}`,
          type: 'video' as const,
          label: sub,
          provider: v?.channel ?? 'YouTube',
          url: v?.url ?? null,
          thumbnail: v?.thumbnail ?? null,
          meta: v?.title ?? '',
        };
      }),
    );
  }

  if (mode === 'mixed' && !videoTasks.length) {
    videoTasks = await Promise.all(
      subtopics.map(async (sub, j) => {
        const vids = await searchVideos(`${sub} tutorial`, apiKey, 1);
        const v = vids[0];
        return {
          id: `${skillId}_sub${j}`,
          type: 'video' as const,
          label: sub,
          provider: v?.channel ?? 'YouTube',
          url: v?.url ?? null,
          thumbnail: v?.thumbnail ?? null,
          meta: v?.title ?? '',
        };
      }),
    );
  }

  tasks.push(...videoTasks);
  tasks.push({
    id: `${skillId}_project`,
    type: 'project',
    label: `Mini-project: build something using ${label} to reinforce this module`,
    provider: 'Self-practice',
    url: null,
    thumbnail: null,
  });
  return tasks;
}

export interface GeneratedResult {
  learningPath: LearningPath;
  skills: Skill[];
  courses: Course[];
  note?: string;
  matchConfidence: number;
}

export async function generateLearningPath(params: GenerationParams): Promise<GeneratedResult> {
  const { goalText, daysAvailable, experienceLevel, freeOnly, resourceMode, youtubeApiKey } = params;
  const matches = parseGoal(goalText);

  const pathId = `gen_${Date.now()}`;
  const skillsOut: Skill[] = [];
  const coursesOut: Course[] = [];
  let stages: PathStage[];
  let domain: string;
  let matchConfidence = 0;
  let note: string | undefined;
  let goalLabel: string;
  const difficulty: LearningPath['difficulty'] =
    (experienceLevel === 'Expert' ? 'Advanced' : experienceLevel) as LearningPath['difficulty'];

  if (!matches.length) {
    // Generic fallback: no taxonomy overlap — still build a real path from
    // YouTube directly, rather than refusing. Every topic gets *something*.
    const topic = cleanTopicPhrase(goalText);
    goalLabel = topic;
    domain = 'custom';
    const subtopics = [
      `${topic} Basics`, `${topic} Core Concepts`,
      `${topic} Intermediate Concepts`, `${topic} Real-World Applications`,
    ];
    const tasks = await buildTasksForSkill('custom', topic, subtopics, [], resourceMode, youtubeApiKey);
    const weeks = daysAvailable ? Math.max(1, Math.ceil(daysAvailable / 7)) : 4;
    stages = [{
      id: 'custom', index: 1, title: topic,
      description: `A generic basics → core → intermediate → applied progression for "${topic}".`,
      status: 'active', durationWeeks: weeks,
      milestones: tasks.map((t) => t.label),
      prerequisites: [],
      why: `"${topic}" isn't in our curated skill catalog yet, so this path is built directly `
        + `from YouTube search rather than our course dataset.`,
      skillIds: ['gen_custom'],
      tasks,
    }];
    skillsOut.push({
      id: 'gen_custom', name: topic, icon: 'Sparkles', current: 0, target: 100,
      targetLabel: params.purpose, level: experienceLevel, mastered: false,
      nextMilestone: tasks[0]?.label ?? 'Get started', domain: 'Custom',
    });
    note = `No exact catalog match for "${goalText}" — generated a generic YouTube-based path `
      + `instead of blocking. Try mentioning a closer domain (e.g. Python, Java, web development, `
      + `data science) for a richer course-backed path.`;
  } else {
    const [goalSkill, score] = matches[0];
    matchConfidence = score;
    const meta = SKILLS[goalSkill];
    goalLabel = meta.label;
    domain = meta.domain;
    const known = inferKnownSkills(experienceLevel, domain);
    let order = shortestLearningOrder(known, goalSkill);
    if (!order.length) order = [goalSkill];

    const n = order.length;
    const daysPerSkill = daysAvailable ? Math.max(1, Math.floor(daysAvailable / n)) : null;
    let dayCursor = 1;

    stages = await Promise.all(
      order.map(async (sid, i) => {
        const smeta = SKILLS[sid];
        const courses = recommendCourses(sid, { level: experienceLevel, freeOnly, topN: 3 });
        const subtopics = getSubtopics(sid, smeta.label);

        const prereqLabels = smeta.prereqs
          .filter((p) => order.includes(p) || known.includes(p))
          .map((p) => SKILLS[p].label);
        let why = `'${smeta.label}' is needed because `
          + (prereqLabels.length ? `it builds on ${prereqLabels.join(', ')}, ` : '')
          + `and is a direct prerequisite on the path to your goal ('${meta.label}').`;
        if (sid === goalSkill) {
          why = `This is your target skill — matched to your goal with ${Math.round(score * 100)}% confidence.`;
        }

        const tasks = await buildTasksForSkill(sid, smeta.label, subtopics, courses, resourceMode, youtubeApiKey);

        const dayStart = dayCursor;
        const dayEnd = daysPerSkill ? dayCursor + daysPerSkill - 1 : null;
        if (daysPerSkill) dayCursor += daysPerSkill;
        const weeks = daysPerSkill ? Math.max(1, Math.ceil(daysPerSkill / 7)) : 2;

        // seed a Course entry per skill (foundational course task) so CoursesPage stays populated
        if (courses.length) {
          const best = courses[0];
          coursesOut.push({
            id: `${pathId}_${sid}_c`,
            title: best.title,
            description: `Recommended ${smeta.label} course to progress toward "${goalText}".`,
            icon: DOMAIN_ICON[smeta.domain] ?? 'BookOpen',
            subject: smeta.label,
            subjectColor: DOMAIN_COLOR[smeta.domain] ?? 'iris',
            resourceType: 'course',
            status: i === 0 ? 'in-progress' : 'recommended',
            progress: 0,
            prerequisites: prereqLabels,
            whyRecommended: why,
            pathId,
            durationHours: best.duration_hours ?? 8,
            bookmarked: false,
          });
        }

        skillsOut.push({
          id: `gen_${sid}`, name: smeta.label, icon: DOMAIN_ICON[smeta.domain] ?? 'Sparkles',
          current: known.includes(sid) ? 70 : 0, target: 100, targetLabel: params.purpose,
          level: experienceLevel, mastered: false,
          nextMilestone: tasks[0]?.label ?? 'Get started', domain: smeta.domain,
        });

        return {
          id: sid, index: i + 1, title: smeta.label,
          description: why,
          status: i === 0 ? 'active' : 'locked',
          durationWeeks: weeks,
          milestones: tasks.map((t) => t.label),
          prerequisites: prereqLabels,
          why,
          skillIds: [`gen_${sid}`],
          tasks,
          dayRange: dayEnd ? `Day ${dayStart}-${dayEnd}` : null,
        };
      }),
    );
  }

  const pathType: PathType =
    params.purpose.toLowerCase().includes('placement') || params.purpose.toLowerCase().includes('job')
      ? 'career'
      : experienceLevel === 'Beginner' ? 'beginner' : 'skill';

  const totalWeeks = stages.reduce((sum, s) => sum + s.durationWeeks, 0);

  const learningPath: LearningPath = {
    id: pathId,
    title: `${goalLabel} — Personalized Path`,
    subtitle: goalText.length > 80 ? goalText.slice(0, 77) + '...' : goalText,
    domain,
    type: pathType,
    difficulty,
    stageCount: stages.length,
    durationWeeks: totalWeeks,
    color: DOMAIN_COLOR[domain] ?? 'iris',
    stages,
    active: true,
  };

  return { learningPath, skills: skillsOut, courses: coursesOut, note, matchConfidence };
}

File: backend/app/ml/path_generator.py
"""
Orchestrator: turns a profile + goal into a full personalized learning path.

Flow:
  1. Resolve a target skill from the chosen career path (if any) and/or the
     free-text goal (TF-IDF match via recommender.parse_goal).
  2. Walk backwards through the prerequisite DAG from the target skill to
     whatever the learner says they already know (shortest_learning_order).
  3. For every skill in that order, rank the best courses (recommender.
     recommend_courses) and expand the skill into its sub-topic curriculum
     (skill_graph.get_subtopics) -- the frontend fills each sub-topic with a
     real YouTube video client-side.
  4. Budget stages across the learner's available days/weeks and return a
     JSON-serializable structure that's persisted verbatim as the
     LearningPath.content JSONB blob.
"""
from __future__ import annotations

import re
import time
from dataclasses import dataclass

from app.ml.recommender import parse_goal, recommend_courses
from app.ml.skill_graph import SKILLS, CAREER_PATH_BY_ID, get_subtopics, shortest_learning_order

LEVEL_TO_STAGE = {"Beginner": 0, "Intermediate": 1, "Advanced": 2, "Expert": 2}
STAGE_ORDER = ["foundation", "core", "advanced"]

DOMAIN_COLOR = {
    "programming": "iris", "data_science": "sage", "web_dev": "amber",
    "cloud": "iris", "cybersecurity": "iris", "business": "amber",
    "design": "sage", "custom": "iris",
}
DOMAIN_ICON = {
    "programming": "Code2", "data_science": "BarChart3", "web_dev": "Atom",
    "cloud": "Cloud", "cybersecurity": "ShieldCheck", "business": "Compass",
    "design": "Palette", "custom": "Sparkles",
}


def _infer_known_skills(experience_level: str, domain_hint: str | None) -> list[str]:
    stage_idx = LEVEL_TO_STAGE.get(experience_level, 0)
    if stage_idx == 0 or not domain_hint:
        return []
    allowed = set(STAGE_ORDER[:stage_idx])
    return [sid for sid, node in SKILLS.items() if node.domain == domain_hint and node.level in allowed]


def _clean_topic_phrase(goal_text: str) -> str:
    filler = re.compile(
        r"\b(i want to|i wanna|learn|study|become a|get a|placement|for|the|please|help me)\b",
        re.IGNORECASE,
    )
    cleaned = re.sub(r"\s+", " ", filler.sub(" ", goal_text)).strip()
    phrase = cleaned or goal_text
    return " ".join(w[:1].upper() + w[1:] if w else w for w in phrase.split())


@dataclass
class GenerationParams:
    goal_text: str
    career_path: str | None
    experience_level: str = "Beginner"
    purpose: str = "personal growth"
    days_available: int | None = None
    free_only: bool = False
    known_skills: list[str] | None = None


def generate_path(params: GenerationParams) -> dict:
    known_skills = list(params.known_skills or [])
    career = CAREER_PATH_BY_ID.get(params.career_path) if params.career_path else None

    goal_skill_id: str | None = None
    match_confidence = 0.0

    if career and career.get("goal_skill"):
        goal_skill_id = career["goal_skill"]
        match_confidence = 1.0  # explicit choice, no ambiguity

    matches = parse_goal(params.goal_text) if params.goal_text else []

    if not goal_skill_id and matches:
        goal_skill_id, match_confidence = matches[0]

    difficulty = "Advanced" if params.experience_level == "Expert" else params.experience_level
    path_id_seed = f"gen_{int(time.time() * 1000)}"

    if not goal_skill_id:
        # Generic fallback -- no taxonomy overlap. Still build something real
        # instead of refusing: a 4-part basics->applied progression that the
        # frontend fills with YouTube search results directly.
        topic = _clean_topic_phrase(params.goal_text or (career["label"] if career else "your goal"))
        subtopics = [
            f"{topic} Basics", f"{topic} Core Concepts",
            f"{topic} Intermediate Concepts", f"{topic} Real-World Applications",
        ]
        weeks = max(1, (params.days_available or 28) // 7)
        stage = {
            "id": "custom", "index": 1, "title": topic,
            "description": f'A generic basics -> core -> intermediate -> applied progression for "{topic}".',
            "status": "active", "duration_weeks": weeks,
            "prerequisites": [],
            "why": f'"{topic}" isn\'t in our curated skill catalog yet, so this stage is built '
                   f"directly from YouTube search rather than our ranked course dataset.",
            "skill_id": "gen_custom",
            "subtopics": subtopics,
            "courses": [],
        }
        return {
            "title": f"{topic} — Personalized Path",
            "subtitle": (params.goal_text or topic)[:120],
            "domain": "custom",
            "difficulty": difficulty,
            "career_path": params.career_path,
            "match_confidence": 0.0,
            "color": DOMAIN_COLOR["custom"],
            "stages": [stage],
            "note": (
                f'No exact catalog match for "{params.goal_text}" — generated a generic '
                f"YouTube-based path instead of blocking. Try mentioning a closer domain "
                f"(Python, Java, web development, data science, cybersecurity...) for a "
                f"richer, course-backed path."
            ),
        }

    meta = SKILLS[goal_skill_id]
    domain = meta.domain
    inferred_known = _infer_known_skills(params.experience_level, domain)
    known_all = sorted(set(known_skills) | set(inferred_known))

    order = shortest_learning_order(known_all, goal_skill_id)
    if not order:
        order = [goal_skill_id]

    n = len(order)
    days_per_skill = max(1, (params.days_available or 0) // n) if params.days_available else None

    stages = []
    for i, sid in enumerate(order):
        smeta = SKILLS[sid]
        courses = recommend_courses(
            sid, level=params.experience_level, free_only=params.free_only, top_n=3,
        )
        subtopics = get_subtopics(sid, smeta.label)

        prereq_labels = [
            SKILLS[p].label for p in smeta.prereqs if p in order or p in known_all
        ]
        if sid == goal_skill_id:
            why = (
                f"This is your target skill for the "
                f"{career['label'] if career else meta.label} path"
                + (f", matched with {round(match_confidence * 100)}% confidence." if match_confidence else ".")
            )
        else:
            builds_on = f"it builds on {', '.join(prereq_labels)} and " if prereq_labels else ""
            why = (
                f"'{smeta.label}' is needed because {builds_on}"
                f"is a direct prerequisite on the path to your goal ('{meta.label}')."
            )

        weeks = max(1, -(-days_per_skill // 7)) if days_per_skill else 2

        stages.append({
            "id": sid,
            "index": i + 1,
            "title": smeta.label,
            "description": why,
            "status": "active" if i == 0 else "locked",
            "duration_weeks": weeks,
            "prerequisites": prereq_labels,
            "why": why,
            "skill_id": sid,
            "domain": smeta.domain,
            "subtopics": subtopics,
            "courses": courses,
        })

    total_weeks = sum(s["duration_weeks"] for s in stages)
    goal_lower = (params.purpose or "").lower()
    path_type = "career" if ("placement" in goal_lower or "job" in goal_lower) else (
        "beginner" if params.experience_level == "Beginner" else "skill"
    )

    return {
        "title": f"{meta.label} — Personalized Path",
        "subtitle": (params.goal_text or meta.label)[:120],
        "domain": domain,
        "type": path_type,
        "difficulty": difficulty,
        "career_path": params.career_path,
        "match_confidence": round(match_confidence, 4),
        "color": DOMAIN_COLOR.get(domain, "iris"),
        "duration_weeks": total_weeks,
        "stages": stages,
        "note": None,
        "_path_id_seed": path_id_seed,
    }
